importlib_metadata/__init__.py

```diff
@@ -398,66 +488,138 @@ class Deprecated:
         stacklevel=pypy_partial(2),
     )
 
-    def __getitem__(self, name):
+    def __getitem__(self, name: _K) -> _V:
         self._warn()
         return super().__getitem__(name)
 
-    def get(self, name, default=None):
+    @overload
+    def get(self, name: _K) -> Optional[_V]:
+        ...  # pragma: no cover
+
+    @overload
+    def get(self, name: _K, default: _T) -> Union[_V, _T]:
+        ...  # pragma: no cover
+
+    def get(self, name: _K, default: Optional[_T] = None) -> Union[_V, _T, None]:
         self._warn()
         return super().get(name, default)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterator[_K]:
         self._warn()
         return super().__iter__()
 
-    def __contains__(self, *args):
+    def __contains__(self, value: object) -> bool:
         self._warn()
-        return super().__contains__(*args)
+        return super().__contains__(value)
 
-    def keys(self):
+    def keys(self) -> KeysView[_K]:
         self._warn()
         return super().keys()
 
-    def values(self):
+    def values(self) -> ValuesView[_V]:
         self._warn()
         return super().values()
 
 
-class SelectableGroups(Deprecated, dict):
+class SelectableGroups(DeprecatedDict[str, EntryPoints]):
     """
     A backward- and forward-compatible result from
     entry_points that fully implements the dict interface.
     """
 
     @classmethod
-    def load(cls, eps):
+    def load(cls, eps: Iterable[EntryPoint]) -> 'SelectableGroups':
         by_group = operator.attrgetter('group')
         ordered = sorted(eps, key=by_group)
         grouped = itertools.groupby(ordered, by_group)
         return cls((group, EntryPoints(eps)) for group, eps in grouped)
 
     @property
-    def _all(self):
+    def _all(self) -> EntryPoints:
         """
         Reconstruct a list of all entrypoints from the groups.
         """
-        groups = super(Deprecated, self).values()
+        groups = super(DeprecatedDict, self).values()
         return EntryPoints(itertools.chain.from_iterable(groups))
 
     @property
-    def groups(self):
+    def groups(self) -> Set[str]:
         return self._all.groups
 
     @property
-    def names(self):
+    def names(self) -> Set[str]:
         """
         for coverage:
         >>> SelectableGroups().names
         set()
         """
         return self._all.names
 
-    def select(self, **params):
+    @overload
+    def select(self) -> 'SelectableGroups':
+        ...  # pragma: no cover
+
+    @overload
```

Comment:
Your explanation is sound and reasonable. Agreed, this interface is confusing because it's trying to support the transition from one return type to another without losing the aptly-named function. Thanks for capturing this type-complicated implementation.